tests/sccache_cargo.rs

```diff
@@ -3,53 +3,101 @@
 //! Any copyright is dedicated to the Public Domain.
 //! http://creativecommons.org/publicdomain/zero/1.0/
 
-#![deny(rust_2018_idioms)]
+use anyhow::{Context, Result};
+use once_cell::sync::Lazy;
+
+use assert_cmd::prelude::*;
+use chrono::Local;
+use predicates::prelude::*;
+use std::ffi::OsString;
+use std::fs;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
 
-#[cfg(all(not(target_os = "windows"), not(target_os = "macos")))]
 #[macro_use]
 extern crate log;
 
-/// Test that building a simple Rust crate with cargo using sccache results in a cache hit
-/// when built a second time.
-#[test]
-#[cfg(all(not(target_os = "windows"), not(target_os = "macos")))]
-fn test_rust_cargo() {
-    test_rust_cargo_cmd("check", &[]);
-    test_rust_cargo_cmd("build", &[]);
+static SCCACHE_BIN: Lazy<PathBuf> = Lazy::new(|| assert_cmd::cargo::cargo_bin("sccache"));
+static CARGO: Lazy<OsString> = Lazy::new(|| std::env::var_os("CARGO").unwrap());
+static CRATE_DIR: Lazy<PathBuf> =
+    Lazy::new(|| Path::new(file!()).parent().unwrap().join("test-crate"));
 
-    #[cfg(feature = "unstable")]
-    test_rust_cargo_cmd("check", &[("RUSTFLAGS", std::ffi::OsStr::new("-Zprofile"))]);
-    #[cfg(feature = "unstable")]
-    test_rust_cargo_cmd("build", &[("RUSTFLAGS", std::ffi::OsStr::new("-Zprofile"))]);
+/// Used as a test setup fixture. The drop implementation cleans up after a _successfull_ test.
+/// It is therefore important that the asserts use `Result` to return errors instead of panicking.
+/// Otherwise the temporary directory will not be cleaned up and the started sccache server process
+/// will not be stopped.
```

Comment:
If we need to do post-test cleanup, we can catch the unwind:
```
let test_result = panic::catch_unwind(|| {
    // do majority of the test
});
// do test cleanup
assert!(test_result.is_ok())
```

This will help avoid a footgun for consumers of `SccacheTest`.

Reply:
For now I just immediately unwrap again after `catch_unwind`, since the `drop` for the `SccacheTest` would have already run then. The unwrapping returns the result of the test, which is maybe a bit unintuitive.

Reply:
Now that I think it again, the `drop` actually runs anyway if `panic=unwind` and will clean up the directory. And if `panic=abort`, then the catch_unwind will also not work so I'm just going to remove it again, since it doesn't help.